### ISSUE TITLE
'Soft' deletion with django-hvad?

### ISSUE THREAD

**Issue by leifurhauks** [KristianOellegaard/django-hvad#263]
I'm trying to implement 'soft' deletion, similar to the functionality in packages like [django-permanent](https://github.com/meteozond/django-permanent), [django-safedelete](https://github.com/makinacorpus/django-safedelete) and others.

If anyone can offer advice on doing this in a project that makes use of `django-hvad` it would be greatly appreciated. In particular, I need soft deletion to cascade (for example, if I soft delete a `Blog`, all `Entries` with foreign keys to that blog should also be soft deleted), and I need for translation models not to be removed from the database when I soft delete a shared model, since that would make the TranslatedModel effectively unrecoverable.

**Comment by spectras** [comment]
I had a look at safedelete, it should not be too hard to make it work with hvad. Maybe it could even work out of the box, I see its factory takes a `manager_superclass` and `queryset_superclass`, which should be able to build upong hvad's `TranslationManager` and `TranslationQueryset`.

It will use the default queryset class however, which means `untranslated()` would not work (it would be there, but ignore the soft deletion mechanism). I guess it would be possible to make some glue code around that.
But safedelete does not do cascade.

The other, django-permanent seems to do cascade, but it digs deeply into Django ORM internals, building queryies by manipulating and monkey-patching the internals. Trying to integrate it with any other third-party module, be it hvad or mptt or anything sounds like calling for trouble.

Depending on the scale, I believe I would just set a boolean field `is_deleted` on the models and filter on it. Basically, that's what django-safedelete does, adding some synctactic sugar on top of it. As for the cascading, I would handle it myself.

Although thinking about it, I wonder if cascading soft delete is even possible in a safe way. I mean, at the SQL level. Maybe by making the `is_deleted` field part of the foreign key, with cascading updates. But that's way out of the scope of Django ORM.

**Comment by leifurhauks** [comment]
Thanks a lot for your advice!

In the end I'm doing basically what you suggest; I have a 'removed' field which tracks whether an object has been soft deleted, and I wrote some code to cascade to related objects that can be soft deleted. I short-circuit the cascading for objects that can't be soft-deleted (like translations), so that they aren't removed from the database during a soft deletion. Of course, I also made it possible to force a regular deletion.
